rainier-core/src/main/scala/com/stripe/rainier/core/Predictor.scala

```diff
@@ -5,66 +5,65 @@ import com.stripe.rainier.compute._
 /**
   * Predictor class, for fitting data with covariates
   */
-trait Predictor[X, Y] extends Likelihood[(X, Y)] {
-  private[core] type Q
-  private[core] type R
-  type P = (Q, R)
-  private[core] def create(q: Q): Distribution.Aux[Y, R]
-  private[core] def xq: Wrapping[X, Q]
+sealed trait Predictor[L, X] { self =>
+  type P
+  protected def xp: Encoder[X, P]
+  protected def create(p: P): L
 
-  lazy val wrapping = new Wrapping[(X, Y), (Q, R)] {
-    def wrap(value: (X, Y)) = {
-      val q = xq.wrap(value._1)
-      val qr = create(q).wrapping
-      (q, qr.wrap(value._2))
-    }
-    def placeholder(seq: Seq[(X, Y)]) = {
-      val qph = xq.placeholder(seq.map(_._1))
-      val z = create(qph.value)
-      qph.zip(z.wrapping.placeholder(seq.map(_._2)))
-    }
-  }
-
-  def logDensity(value: P) =
-    create(value._1).logDensity(value._2)
+  def fit[Y](values: Seq[(X, Y)])(
+      implicit lh: ToLikelihood[L, Y]): RandomVariable[Predictor[L, X]] =
+    Predictor
+      .likelihood[L, X, Y](this)
+      .fit(values)
+      .map { _ =>
+        this
+      }
 
-  def predict(x: X): Generator[Y] =
-    create(xq.wrap(x)).generator
+  def predict[Y](x: X)(implicit gen: ToGenerator[L, Y]): Generator[Y] =
+    gen(create(xp.wrap(x)))
 
-  def predict(seq: Seq[X]): Generator[Seq[(X, Y)]] =
+  def predict[Y](seq: Seq[X])(
+      implicit gen: ToGenerator[L, Y]): Generator[Seq[(X, Y)]] =
     Generator.traverse(seq.map { x =>
       predict(x).map { y =>
         (x, y)
       }
     })
 }
 
-/**
-  * Predictor object, for fitting data with covariates
-  */
-object Predictor {
-  def from[X, Y, A, B](fn: A => Distribution.Aux[Y, B])(
-      implicit xa: Wrapping[X, A]): Predictor[X, Y] =
-    new Predictor[X, Y] {
-      type Q = A
-      type R = B
-
-      val xq = xa
-      def create(q: Q) = fn(q)
+object Predictor extends LikelihoodMaker {
+  def likelihood[L, X, Y](pred: Predictor[L, X])(
+      implicit lh: ToLikelihood[L, Y]): Likelihood[(X, Y)] = {
+    val (p, v) = pred.xp.create(Nil)
+    val l = pred.create(p)
+    val inner = lh(l)
+    new Likelihood[(X, Y)] {
+      val real = inner.real
+      val placeholders = v ++ inner.placeholders
+      def extract(t: (X, Y)) =
+        pred.xp.extract(t._1, Nil) ++ inner.extract(t._2)
     }
+  }
 
-  def fromInt[Y, B](fn: Real => Distribution.Aux[Y, B]): Predictor[Int, Y] =
-    from[Int, Y, Real, B](fn)
+  class Maker[X, A](xa: Encoder[X, A]) {
+    def apply[B](fn: A => B): Predictor[B, X] =
+      new Predictor[B, X] {
+        type P = A
+        val xp = xa
+        def create(p: P) = fn(p)
+      }
+  }
 
-  def fromIntPair[Y, B](
-      fn: ((Real, Real)) => Distribution.Aux[Y, B]): Predictor[(Int, Int), Y] =
-    from[(Int, Int), Y, (Real, Real), B](fn)
+  type M[X, A] = Maker[X, A]
```

Comment:
I don't this gets used.

Reply:
It does so that both `Predictor` and `Likelihood` can inherit from `LikelihoodMaker`. (I agree with you that this is all crazy convoluted).